streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StickyTaskAssignor.java

```diff
@@ -0,0 +1,280 @@
+            for (final TaskId taskId : taskIds) {
+                final Set<ID> ids = findClientsWithoutAssignedTask(taskId);
+                if (ids.isEmpty()) {
+                    log.warn("Unable to assign replica for task [{}]", taskId);
+                    continue;
+                }
+                assign(taskId, ids, false);
+            }
+        }
+    }
+
+    private void assignActive() {
+        final Set<TaskId> previouslyAssignedTaskIds = new HashSet<>(previousActiveTaskAssignment.keySet());
+        previouslyAssignedTaskIds.addAll(previousStandbyTaskAssignment.keySet());
+        previouslyAssignedTaskIds.retainAll(taskIds);
+
+        // assign previously assigned tasks first
+        for (final TaskId taskId : previouslyAssignedTaskIds) {
+            assign(taskId, clients.keySet(), true);
+        }
+
+        final Set<TaskId> newTasks  = new HashSet<>(taskIds);
+        newTasks.removeAll(previouslyAssignedTaskIds);
+
+        for (final TaskId taskId : newTasks) {
+            assign(taskId, clients.keySet(), true);
+        }
+    }
+
+    private void assign(final TaskId taskId, final Set<ID> clientsWithin, final boolean active) {
+        final ClientState<TaskId> client = findClient(taskId, clientsWithin);
+        taskPairs.addPairs(taskId, client.assignedTasks());
+        client.assign(taskId, active);
+    }
+
+    private Set<ID> findClientsWithoutAssignedTask(final TaskId taskId) {
+        final Set<ID> clientIds = new HashSet<>();
+        for (final Map.Entry<ID, ClientState<TaskId>> client : clients.entrySet()) {
+            if (!client.getValue().hasAssignedTask(taskId)) {
+                clientIds.add(client.getKey());
+            }
+        }
+        return clientIds;
+    }
+
+
+    private ClientState<TaskId> findClient(final TaskId taskId,
+                                           final Set<ID> clientsWithin) {
+        // optimize the case where there is only 1 id to search within.
+        if (clientsWithin.size() == 1) {
+            return clients.get(clientsWithin.iterator().next());
+        }
+
+        final ClientState<TaskId> previous = findClientsWithPreviousAssignedTask(taskId, clientsWithin);
+        if (previous == null) {
+            return leastLoaded(taskId, clientsWithin);
+        }
+
+        if (shouldBalanceLoad(previous)) {
+            final ClientState<TaskId> standby = findLeastLoadedClientWithPreviousStandByTask(taskId, clientsWithin);
+            if (standby == null
+                    || shouldBalanceLoad(standby)) {
+                return leastLoaded(taskId, clientsWithin);
+            }
+            return standby;
+        }
+
+        return previous;
+    }
+
+    private boolean shouldBalanceLoad(final ClientState<TaskId> client) {
+        return !hasNewTasks
+                && client.reachedCapacity()
+                && availableCapacity <= taskIds.size()
+                && hasClientsWithZeroTasks();
+    }
```

Comment:
I don't get this condition:
I understand (1) `!hasNewTasks` (that is only true, if we scale out and not loose any existing client, right?) and I understand (2) `client.reachedCapacity()` (no reason to assign somewhere else I client has spare capacity).

But the last two conditions puzzle me.

(3) Let's say I have 1 client with capacity 1 and overall 2 tasks, both assigned to it (ie, `taskIds.size() == 2`). I add a new client with capacity 2. Now `availableCapacity == 3` and `taskIds.size()` is still `2`. Thus `availableCapacity <= taskIds.size()` (`3 <= 2`) is `false`, and no load rebalancing would happen. However, this would result in an overload of the original client (2 tasks assigned with capacity 1), while the new client is under utilized (0 tasks assigned with capacity 2). What do I miss?

(4) Let's say I have 1 client with capacity 1 and overall 3 tasks, all assigned to it (ie, `taskIds.size() == 3`).  I add a new client with capacity 2. Condition (3) would be met because (`3 <= 3`) so one task would be re-assigned from it's original client to the new client. But not the second tasks, because after the first client got a task assigned there is no client with zero tasks left. Thus the original client would still be overloaded (2 tasks assigned with capacity 1) while the new client is under utilized (1 tasks assigned with capacity 2). What do I miss?